### ISSUE TITLE
Debug build variant doesn't pass correct flags to GCC with Makefiles

### ISSUE THREAD

**Issue by jasonbeach** [microsoft/vscode-cmake-tools#1031]
### Brief Issue Summary
Debug build variant doesn't pass correct flags to GCC.  

### Apparent Behavior:
Ubuntu 16.04, cmake 3.14.6, vscode  1.41.1, c/c++ extension 0.26.3, cmake tools 1.2.3
building within docker remote dev container. Makefile build generator

From a completely fresh build (i.e. `rm -rf` build)
1. I configure the project via clicking the "Click to select build variant" button on the status ribbon at the bottom of the screen and then select `Debug` in the pop up that appears. 
2. `cmake` configures without errors
3. I push the `build` button at the bottom (kit had been previously selected.)
4. Building commences but it becomes apparent that not all the flags given in `CMakeLists.txt` were passed to the compiler.  Once the build fails, in the terminal I manually start the build via `make VERBOSE=1` and looking at the GCC command, sure enough the flags are not being passed to GCC. 

If I nuke the build directory and manually configure (i.e. `cmake -DCMAKE_BUILD_TYPE=Debug ..`) then everything works as it should.  This only appears to happen when selecting the `DEBUG` build variant.  Repeating the above steps, but selecting `Release` instead of `Debug` works as expected.

In my case I was trying to declare some directories as `SYSTEM` (i.e. use `-isystem` instead of `-I` via `target_include_directories`), however I observed similar behavior with a co-worker yesterday.  She was trying to debug a program and kept getting an error along the lines of symbols not found.  When manually starting the build via `make VERBOSE=1` I noted the `-g` option was not being passed to GCC in spite of having the `Debug` variant selected. In that case too, nuking the build directory and manually running `cmake` from the terminal made things run as they should.

### Expected:
Correct flags passed to GCC

### CMake Tools Log
﻿I've slightly snipped the following log to omit some of the messages I print from cmake.
```
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:26.024Z [debug] [extension] [9211] cmake.setVariant started
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.424Z [debug] [main] Active build variant changed
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.426Z [debug] [main] Run configure 
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.427Z [debug] [main] Saving open files before configure/build
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.456Z [debug] [driver] Setting new variant , Disable optimizations - include debug information.
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.457Z [debug] [driver] Run _refreshExpansions
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.457Z [debug] [cms-driver] Run doRefreshExpansions
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.457Z [debug] [driver] Run _refreshExpansions cb
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.533Z [debug] [driver] Start configure 
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.533Z [debug] [driver] Runnnig pre-configure checks and steps
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.536Z [debug] [driver] Using compilers in x86-gcc740 for configure
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.538Z [debug] [driver] Run _refreshExpansions
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.539Z [debug] [cms-driver] Run doRefreshExpansions
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.539Z [debug] [driver] Run _refreshExpansions cb
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:30.831Z [info] [cmake] The C compiler identification is GNU 7.4.0
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.030Z [info] [cmake] The CXX compiler identification is GNU 7.4.0
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.053Z [info] [cmake] Check for working C compiler: /usr/bin/gcc-7
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.188Z [info] [cmake] Check for working C compiler: /usr/bin/gcc-7 -- works
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.194Z [info] [cmake] Detecting C compiler ABI info
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.346Z [info] [cmake] Detecting C compiler ABI info - done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.377Z [info] [cmake] Detecting C compile features
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.604Z [info] [cmake] Detecting C compile features - done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.612Z [info] [cmake] Check for working CXX compiler: /usr/bin/g++-7
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.704Z [info] [cmake] Check for working CXX compiler: /usr/bin/g++-7 -- works
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.706Z [info] [cmake] Detecting CXX compiler ABI info
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.784Z [info] [cmake] Detecting CXX compiler ABI info - done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:31.802Z [info] [cmake] Detecting CXX compile features
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.412Z [info] [cmake] Detecting CXX compile features - done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.413Z [info] [cmake] CMAKE: /usr/local/bin/cmake Version: 3.14.6
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.413Z [info] [cmake] CMAKE_INSTALL_PREFIX not set in cache or on command line. Setting to in-source deploy folder.
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.414Z [info] [cmake] CMAKE_INSTALL_PREFIX: /workspaces/focus/deploy
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.864Z [info] [cmake] Found Protobuf: /usr/local/lib/libprotobuf.so;-lpthread (found suitable version "2.6.1", minimum required is "2.6.1") 
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.869Z [info] [cmake] Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.870Z [info] [cmake] Checking for module 'opencv'
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:32.891Z [info] [cmake]   Found opencv, version 3.2.0
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.102Z [info] [cmake] VEH application configured
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.143Z [info] [cmake] PROTOC executable is /usr/local/bin/protoc
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.143Z [info] [cmake] proto library configured
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.470Z [info] [cmake] Found Doxygen: /usr/bin/doxygen (found version "1.8.13") found components:  doxygen dot 
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.590Z [info] [cmake] doc_doxygen target configured
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.611Z [info] [cmake] Configuring done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.868Z [info] [cmake] Generating done
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.978Z [debug] [cache] Reading CMake cache file /workspaces/focus/build/x86-gcc740-Debug/CMakeCache.txt
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:33.988Z [debug] [cache] Parsing CMake cache string
console.ts:137 [Extension Host] [CMakeTools] 2020-01-29T19:56:34.075Z [debug] [extension] [9211] cmake.setVariant finished (returned true)

```

**Comment by andreeis** [comment]
Can you share the CMakeLists.txt to see how the variant specific switches are defined? And also the verbose log for the debug and release build.

**Comment by jasonbeach** [comment]
When selected release the`-O3` flags get set as well as the expected `-isystem` flags.  In my `CMakeLists.txt` I don't define any build variant sets of flags.  I only have:
```
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

# Be super strict about compiler warnings
OPTION (ENABLE_WARNINGS_AS_ERRORS "Compile with -Werror in gcc ....... good luck" ON)

add_compile_options(-Wall -Wextra -pedantic)
```
and then in nested CMakeLists.txt for each target I check whether or not `ENABLE_WARNINGS_AS_ERRORS` is set and add the flag accordingly, i.e.

```
IF (ENABLE_WARNINGS_AS_ERRORS)
    target_compile_options(configuration_reader PRIVATE -Werror)
ENDIF()
```

This appears to be sort of a heisenbug--sometimes it consistently works as it should and other times it consistently doesn't and I haven't figured out what the delta is between the two. 

In the top level of my project I have a `build` directory.  In my repository specific `settings.json` I have 
```
"cmake.buildDirectory": "${workspaceRoot}/build/${buildKit}-${buildType}",
```
and in my local `cmake-kits.json` I have this kit defined:
```
  {
    "name": "x86-gcc740",
    "compilers": {
      "C": "/usr/bin/gcc-7",
      "CXX": "/usr/bin/g++-7"
    }
  },
```
(which I recognize is the same as the one that the system automatically finds, but has a less egregious name for creating the build directory as above).  The goal being that within the `build` directory there is a nested directory for each type of build, i.e. x86 debug, x86 release, aarch64 debug, etc.)  For purposes of this bug though, I'm not using the cross compiler. 

 What seems to reliably trigger this is:
1. have the kit (i.e. `x86-gcc740`) selected from previous builds
2. from the terminal nuke all of the nested build directories (i.e. `rm -rf *` from within the `build` directory)
3. Click the `CMake: Debug: Ready` button on the ribbon and then select `Debug`

At this point you don't even need to start the build--you can look at the CMakeCache.txt and see the flags are missing. 

I have an example CMakeCache.txt from a manual configure that completed successfully and one that resulted from the process above.  I would rather not share them publically--is there a way I can share them privately?

How do I get the verbose log?

**Comment by jasonbeach** [comment]
So with a little more testing it appears that if I reselect the kit in between nuking the build directory and selecting the build variant things work as they should.  i.e. this works:

1. from the terminal nuke all of the nested build directories (i.e. rm -rf * from within the build directory)
2. click the kit selection button at the bottom (even though it says x86-gcc740 is the active kit), and re-select the kit.
3. Click the CMake: Debug: Ready button on the ribbon and then select Debug

whereas the above steps don't.  One potential clue is that in instances where it works, the start of the cmake output has:

```
[cmake] The C compiler identification is GNU 7.4.0
[cmake] The CXX compiler identification is GNU 7.4.0
[cmake] Check for working C compiler: /usr/bin/gcc-7
[cmake] Check for working C compiler: /usr/bin/gcc-7 -- works
...
```
and when it doesn't work it has:
```
[cmake] The C compiler identification is unknown
[cmake] The CXX compiler identification is unknown
[cmake] Check for working C compiler: /usr/bin/gcc-7
[cmake] Check for working C compiler: /usr/bin/gcc-7 -- works
...
```

**Comment by jasonbeach** [comment]
Are you still needing more info on this?

**Comment by andreeis** [comment]
Not for now, thank you. We are going to investigate and fix this for the 1.4 release.

**Comment by DDevine** [comment]
I've stumbled across this issue, I've tried some of the workarounds @jasonbeach suggested (deleting build directly, resetting kit and variant) but nothing seems to result in the debug flags being added to the "Configured C_Flags". 

I don't see any issue with "...compiler identification is unknown" in my case. 

I'm on version v1.7.2.
